source/time_tables/departure_boards.cpp

```diff
@@ -270,11 +346,75 @@ void departure_board(PbCreator& pb_creator, const std::string& request,
+           calendar_id, depth);
 
     pb_creator.make_paginate(total_result, start_page, count, std::max(pb_creator.departure_boards_size(),
                                                                        pb_creator.stop_schedules_size()));
 }
+
+DateTime first_or_last_request_datetime(const datetime_type type,
+                                        const routing::datetime_stop_time& first_stop_time,
+                                        const uint32_t opening_time)
+{
+    // First date time case.
+    // The request date time is equal to the opening date time -1 sec.
+    if (type == datetime_type::first) {
+        return DateTimeUtils::date(first_stop_time.first)*DateTimeUtils::SECONDS_PER_DAY + opening_time - 1;
+    }
+    // Last date time case.
+    else {
+        // The request date time is equal to the opening date time -1 sec.
+        if (first_stop_time.second->departure_time <= opening_time) {
+            return (DateTimeUtils::date(first_stop_time.first))*DateTimeUtils::SECONDS_PER_DAY + opening_time - 1;
+        }
+        // The request date time is equal to the opening date time the day after -1 sec.
+        else {
+            return (DateTimeUtils::date(first_stop_time.first) + 1)*DateTimeUtils::SECONDS_PER_DAY + opening_time - 1;
+        }
+    }
 }
+
+boost::optional<routing::datetime_stop_time>
+get_one_stop_time(const datetime_type type,
+                  const routing::datetime_stop_time& first_stop_time,
+                  const pt::time_duration& opening_time,
+                  const std::vector<routing::JppIdx>& journey_pattern_points,
+                  const DateTime max_datetime,
+                  const type::Data& data,
+                  const type::RTLevel rt_level,
+                  const size_t utc_offset)
+{
+    DateTime max_dt = max_datetime;
+    if (type == datetime_type::last) {
+        max_dt = 0; // clockwise == false
+    }
+    // if the opening_time is 03:00(local_time) and the utc_offset is 5 hours, we want 22:00(UTC) as opening time
+    int local_opening_time = math_mod(opening_time.total_seconds() - utc_offset, DateTimeUtils::SECONDS_PER_DAY);
+    assert(local_opening_time > 0);
+    DateTime new_current_time = first_or_last_request_datetime(type,
+                                                               first_stop_time,
+                                                               static_cast<uint32_t>(local_opening_time));
+
+    auto stop_times = routing::get_stop_times(routing::StopEvent::pick_up,
+                                         journey_pattern_points,
+                                         new_current_time,
+                                         max_dt,
+                                         1,
+                                         data,
+                                         rt_level);
+    if (!stop_times.empty()) {
+        return stop_times[0];
+    } else {
+        return boost::optional<routing::datetime_stop_time>{};
```

Comment:
An empty optional can simply be returned with: `boost::none`

Reply:
:+1: